### ISSUE TITLE
raise ignored in favor of undefined method

### ISSUE THREAD

**Issue by Exilor** [manastech/crystal#398]
```ruby
class Foo(T)
  def initialize(@bar : T)
  end

  def numeric?
    T.responds_to?(:cast)
  end

  def baz
    raise Exception.new "not numeric" unless numeric?
    @bar + 5
  end
end

Foo.new(5).baz # => 10
Foo.new('5').baz # undefined method '+' for Char
```
I expected the second example to raise. Is it a bug or am I missing something here? I'm positive that #numeric? works as I intended.
I also tried:
```ruby
def baz
  if @bar.is_a?(Number)
    @bar + 5
  else
    raise Exception.new "not numeric"
  end
end
```
and I get the same error.

**Comment by asterite** [comment]
The compiler doesn't do a very smart deep type flow analysis. The compiler doesn't deduce anything from the `numeric?` call.

Some questions:

1. Do you want to allow T to only be used by types that extend Number? If that's the case, we are missing that future (something like `class Foo(T : Number)`).
2. Why do you want to raise a runtime error when you can raise a compile time error?

If you want a runtime exception if T is not a Number, you can do this:

```crystal
class Foo(T)
  def initialize(@bar : T)
  end

  def baz
    bar = @bar
    if bar.is_a?(Number)
      bar + 5
    else
      raise "not numeric"
    end
  end
end

puts Foo.new(5).baz # => 10
puts Foo.new('5').baz # raises "not numeric"
```

Basically, in Crystal the type flow only applies to local variables. So you must first assign the instance variable to a local variable. Then if you do `is_a?(...)` the compiler will know that inside the `if` the variable will be of that type. Anything more complex than that and the compiler won't understand it.

You can also do this:

```crystal
class Foo(T)
  def initialize(@bar : T)
  end

  def baz
    bar = @bar
    if bar.responds_to?(:+)
      bar + 5
    else
      raise "not numeric"
    end
  end
end

puts Foo.new(5).baz # => 10
puts Foo.new('5').baz # raises "not numeric"
```

**Comment by Exilor** [comment]
1. I wanted to make a generic class of any type that had some methods only usable to instances where #numeric? returns true.
2. I actually don't know how to raise a compile instead of runtime error in this particular case.


But essentially it boils down to me erroneously assuming that instance variables would behave under the same type rules as local variables, as you explained.
